### DISCUSSION TITLE
Return both data and errors from a resolver

### DISCUSSION THREAD

**Discussion by Eetami** [graphql-java-kickstart/graphql-spring-boot#389] [Q&A]
I have a schema akin to the following
```gql
# YAML formatted in String
scalar YAML
type DocumentWrapper {
  id: ID!
  name: String
  document(validate: Boolean = false): YAML
}
type Query {
  # Get a single document wrapper entity
  documentWrapper(id: ID!): DocumentWrapper
}
```

And resolver classes for both Query and DocumentWrapper

```java
@Component
public class Query implements GraphQLQueryResolver {
  @Autowired
  private DocumentWrapperRepository documentWrapperRepository;

  public DocumentWrapper getDocumentWrapper(UUID id) {
    return documentWrapperRepository.findById(id).get();
  }
}
```

```java
@Component
public class DocumentWrapperResolver implements GraphQLResolver<DocumentWrapper> {
  @Autowired
  private DocumentValidator documentValidator;

  public DataFetcherResult<YAML> document(DocumentWrapper parent, boolean validate) {
    DataFetcherResult.Builder<YAML> builder = DataFetcherResult.<YAML>newResult().data(parent.getDocument());
    if (validate) {
      List<GraphQLError> errors = documentValidator.validate(parent.getDocument());
      builder.errors(errors).mapRelativeErrors(true);
    }
    return builder.build();
  }
}
```

The DocumentValidator returns a list of errors that are of type DocumentValidationError

```java
public class DocumentValidationError extends GraphQLException implements GraphQLError{
  @Override
  public String getMessage() {
    // Return the root problem from cause
  }

  @Override
  public List<SourceLocation> getLocations() {
    return Collections.emptyList(); // Parent source location
  }

  @Override
  public ErrorClassification getErrorType() {
    return YAMLErrorType.DocumentValidationError;
  }

  @Override
  public List<Object> getPath() {
    return Collections.emptyList(); // Parent path
  }

  @Override
  public Map<String, Object> getExtensions() {
    // Return my extensions as map
    Map<String, Object> extensions = new LinkedHashMap<>();
    // Fill with details about  the validation error
    return extensions;
  }
}
```

With this setup I can get both the data and errors in the query response, BUT in the errors I get only the message, not extensions, locations or path. How can I get the extensions, locations and path showing in the query response in case there are validation errors?

**Comment by oliemansm** [comment]
Take a look at the original [GraphQLErrorBuilder](https://github.com/graphql-java/graphql-java/blob/5c242a1a660218519d96353d2e1334e65434f747/src/main/java/graphql/GraphqlErrorBuilder.java) for inspiration. You have to add the `DataFetchingEnvironment` as the last argument of your resolver method and pass it along to your validator.

**Comment by Eetami** [comment]
> Take a look at the original [GraphQLErrorBuilder](https://github.com/graphql-java/graphql-java/blob/5c242a1a660218519d96353d2e1334e65434f747/src/main/java/graphql/GraphqlErrorBuilder.java) for inspiration. You have to add the `DataFetchingEnvironment` as the last argument of your resolver method and pass it along to your validator.

Unfortunately this had no effect: still no extensions, locations or path in the errors :/

**Comment by oliemansm** [comment]
I don't know what you actually did now, but I hope you at least removed your getters for locations, path and extensions from your custom implementation. Seeing that you return empty or null there which explains why they are not present in the response.

**Comment by Eetami** [comment]
Well the validator now returns a list of `GraphQLErrorImpl` instances

```java
GraphqlErrorBuilder.newError(environment)
  .errorType(YAMLErrorType.DocumentValidationError)
  .message(validationError.getRootProblem())
  .extensions(validationError.getExtensions())
  .build();
```

So yes, the returned `GraphQLError`s should have path and locations set from the `environment`.

**Comment by Eetami** [comment]
I even tried return again instances of `DocumentValidationError` and setting the locations and paths for them from the environment.

To be frank, I don't really care that locations and path are missing, I just need the extensions to show up, as they will offer more detailed messages where and why the document validation failed.

**Comment by oliemansm** [comment]
Try enabling `graphql.servlet.exception-handlers-enabled=true`. Shouldn't matter, but that's the only thing that quickly comes to mind to try out. Anything else requires trying to reproduce, debugging and digging through the code which will take longer.

**Comment by Eetami** [comment]
I have it enabled already :/ And disabling it causes some misery with GraphiQL (can't get any response).

**Comment by oliemansm** [comment]
@Eetami Did you ever get this to work or still a problem?

**Comment by Eetami** [comment]
Hey @oliemansm I think I found out the cause. After some testing I could not repro this issue in my sandbox project. Though I noticed that in the erronous project we had property `graphql.servlet.exception-handlers-enabled` set to `true`. After disabling it, I was able to see all details for some throwaway errors.

Though now I have a problem that when attempting to use my own error class, the response from GraphQL API is empty.

edit: I was able to fix the issue. The main problem was that GraphQL servlet swallowed a JSON serialization exception, so I had to increase log level to debug to see it. After that it was smooth sailing.

I'll add a link to the swallowing issue for future reference: https://github.com/graphql-java-kickstart/graphql-java-servlet/issues/250

Perhaps the exception should be propagated to Spring exception handlers, so that the developer can see the stack trace even when log level is not debug.

**Comment by spac-valentin** [comment]
Hi, I think I have a similar issue. From one of my resolvers I am returning:
```
return DataFetcherResult.<MyData>newResult()
		.errors(new MyGraphQLError("Return me", extensions)))
	        .data(data)
	        .build();
```

With `graphql.servlet.exception-handlers-enabled=true` I see only the message 
``` "errors": [{"message": "Return me"]} ``` 

while if I set `exception-handlers-enabled=false`, I can also see the extensions 

```"errors": [{  "message": "Return me", "extensions": {"errorCode": "one"}}```.

This behaviour is from `GraphQLErrorFromExceptionHandler`, which creates a `GenericGraphQLError` if can't extract the exception (and `GraphQLError` type is not taken into account)

```
  private Optional<Throwable> extractException(GraphQLError error) {
    if (error instanceof ExceptionWhileDataFetching) {
      return Optional.of(((ExceptionWhileDataFetching) error).getException());
    } else if (error instanceof SerializationError) {
      return Optional.of(((SerializationError) error).getException());
    } else if (error instanceof GraphQLException) {
      return Optional.of((GraphQLException) error);
    }
    return Optional.empty();
  }
```

Is this intended or it's a bug?

**Reply by oliemansm** [reply]
I'd say that's a bug, thanks.

**Reply by oliemansm** [reply]
Really no clue why it was not just returning the original `GraphQLError` argument here in the first place. PR #606
